pkg/tagset/slice_builder.go

```diff
@@ -0,0 +1,186 @@
+// tag's level).
+func (bldr *SliceBuilder) Add(level int, tag string) {
+	h := murmur3.StringSum64(tag)
+	if _, seen := bldr.seen[h]; seen {
+		return
+	}
+	bldr.seen[h] = struct{}{}
+
+	newLen := len(bldr.tags) + 1
+	insertAt := bldr.offsets[level+1]
+
+	var newTags []string
+	var newHashes []uint64
+
+	// reallocate the storage if there is not enough room in the existing
+	// array, and copy the existing data into the new array
+	if cap(bldr.tags) < newLen {
+		newTags = make([]string, newLen, newLen*2)
+		newHashes = make([]uint64, newLen, newLen*2)
+
+		copy(newTags[:insertAt], bldr.tags[:insertAt])
+		copy(newTags[insertAt+1:], bldr.tags[insertAt:])
+
+		copy(newHashes[:insertAt], bldr.hashes[:insertAt])
+		copy(newHashes[insertAt+1:], bldr.hashes[insertAt:])
+	} else {
+		newTags = bldr.tags[:newLen]
+		newHashes = bldr.hashes[:newLen]
+
+		// make space for the new element with the minimum number of copies,
+		// moving the first element of each level's slice to the end of that
+		// slice
+		for l := bldr.levels - 1; l > level; l-- {
+			s := bldr.offsets[l]
+			d := bldr.offsets[l+1]
+			newTags[d], newHashes[d] = newTags[s], newHashes[s]
+		}
+	}
+
+	newTags[insertAt] = tag
+	newHashes[insertAt] = h
+
+	bldr.tags = newTags
+	bldr.hashes = newHashes
+
+	bldr.levelHashes[level] ^= h
+	for l := level + 1; l < bldr.levels+1; l++ {
+		bldr.offsets[l]++
+	}
+}
+
+// AddKV adds the tag "k:v" to the builder at the given level. If the tag is
+// already in the builder, it is not added again (regardless of the existing
+// tag's level).
+func (bldr *SliceBuilder) AddKV(level int, k, v string) {
+	tag := k + ":" + v
+	bldr.Add(level, tag)
+}
+
+// FreezeSlice "freezes" the builder and returns the requested slice of levels.
+// The Add methods cannot be called after freezing.
+func (bldr *SliceBuilder) FreezeSlice(a, b int) *Tags {
+	// free unnecessary memory
+	bldr.seen = nil
+
+	hash := uint64(0)
+	for _, lh := range bldr.levelHashes[a:b] {
+		hash ^= lh
+	}
+
+	// Whether to use the byTagsetHashCache here is a judgement call. The
+	// downside is that it may end up caching a Tags that refers to a slice
+	// larger than necessary. It's also an additional hash-table lookup when
+	// we already have all of the data necessary to create a new Tags instance.
+	// That said, in cases where only low-cardinality tags are used, it may
+	// result in significant memory savings.
+	//
+	// Another option to explore is to cache the entire built set of tags, and
+	// query that cached value on the first call to FreezeSlice, borrowing the
+	// cached tags and hashes values if there is a match.
+
+	return bldr.factory.getCachedTags(byTagsetHashCache, hash, func() *Tags {
+		tags := bldr.tags[bldr.offsets[a]:bldr.offsets[b]]
+		hashes := bldr.hashes[bldr.offsets[a]:bldr.offsets[b]]
+		return &Tags{tags, hashes, hash}
+	})
+}
+
+// Close closes the builder, freeing its resources.
+func (bldr *SliceBuilder) Close() {
```

Comment:
Can `Freeze` and `Close` be merged?

Reply:
In `Builder`, they are.  But not in this case (`SliceBuilder`), since multiple slices can be frozen before the builder is closed.